### PR TITLE
[Idea] 0042 MetaMask Snap for the XRP Ledger

### PR DESCRIPTION
**Pull request by intelliot**
| [View Proposal Here](https://github.com/XRPLBounties/Proposals/pull/41/files?short_path=78b2d60#diff-78b2d60689ebb8873858b71d47ac5ce2935f09e9827512d87f79173b6ebe4074) |
| - |

### CODE REVIEW

File: bounties/MetaMask Snap for the XRP Ledger.md [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+| - | ----------- | ------- | ------------------------ |
+| 1 | Create Account | Implement a Snap that securely generates and stores a new keypair and displays the associated XRPL address to the user. Make it easy for users to copy the address to share or use in other apps. | $1,000 |
+| 2 | Send XRP | Enable users to send XRP. The Snap should sign and submit the transaction to a server of the user's choice. For convenience, users can select from one of the [well-known public servers](https://xrpl.org/public-servers.html). | $2,000 |
+| 3 | View Transaction History | Show transaction history in a human-readable UI. | $10,000 |
+| 4 | Set Trust Line | Enable users to [create or modify a trust line](https://xrpl.org/trustset.html). | $2,000 |
+| 5 | Sign Anything | Enable users to sign any transaction or message. | $3,000 |
+| 6 | Sign and Submit Transaction | Enable users to sign and submit any transaction. | $3,000 |
+| 7 | Dapp Integration | Offer APIs for dapps to interact with the Snap. For an example, see the [API to integrate BitcoinSnap into your dapp](https://github.com/KeystoneHQ/btcsnap/tree/master/packages/snap). | $4,000 |
```
Comment (jonathanlei) [inline bounties/MetaMask Snap for the XRP Ledger.md:33]: IMO this should be worth more, if API documentation is included

Comment (tinytb) [inline bounties/MetaMask Snap for the XRP Ledger.md:33]:
```suggestion
| 7 | Dapp Integration | Offer APIs for dapps to interact with the Snap. For an example, see the [API to integrate BitcoinSnap into your dapp](https://github.com/KeystoneHQ/btcsnap/tree/master/packages/snap). Additionally, write API documentation, with snippets of example code. | $8,000 |
```

Comment (intelliot) [reply bounties/MetaMask Snap for the XRP Ledger.md:33]:
```suggestion
| 7 | Dapp Integration | Offer APIs for dapps to interact with the Snap, with API documentation and snippets of example code. As a reference, see the [API to integrate BitcoinSnap into your dapp](https://github.com/KeystoneHQ/btcsnap/tree/master/packages/snap). | $8,000 |
```

**Review by free2ride19 — COMMENTED** [review]
Open bounty source - approve
@free2ride19

**Comment by free2ride19** [comment]
Create a UI / library for event organizer to mint and distribute Proof of Attendance NFTs (with event metadata - <date, description, cover art...> embedded in memo field)

File: bounties/MetaMask Snap for the XRP Ledger.md [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+| 4 | Set Trust Line | Enable users to [create or modify a trust line](https://xrpl.org/trustset.html). | $2,000 |
+| 5 | Sign Anything | Enable users to sign any transaction or message. | $3,000 |
+| 6 | Sign and Submit Transaction | Enable users to sign and submit any transaction. | $3,000 |
+| 7 | Dapp Integration | Offer APIs for dapps to interact with the Snap, with API documentation and snippets of example code. As a reference, see the [API to integrate BitcoinSnap into your dapp](https://github.com/KeystoneHQ/btcsnap/tree/master/packages/snap). | $8,000 |
+| 8 | Sample App | Build a sample app to showcase how to interact with the Dapp Integration milestone. | $5,000 |
+| 9 | View NFTs | Enable users to see their XLS-20 NFTs directly within MetaMask. | $5,000 |
+| 10 | Mint NFTs | Enable users to mint or sell a new NFT. | $7,000 |
+| 11 | Ethereum Feature Parity | Add features similar or comparable to those available when using MetaMask with Ethereum. | Up to $50,000 |
+| 12 | Xumm Feature Parity | Add features similar or comparable to those offered by [Xumm](https://xumm.app/). | Up to $50,000 |
```
Comment (mvadari) [inline bounties/MetaMask Snap for the XRP Ledger.md:38]:
```suggestion
| 1 | Create Account | Implement a Snap that securely generates and stores a new keypair and displays the associated XRPL address to the user. Make it easy for users to copy the address to share or use in other apps. | $1,000 |
| 2 | Send XRP | Enable users to send XRP. The Snap should sign and submit the transaction to a server of the user's choice. For convenience, users can select from one of the [well-known public servers](https://xrpl.org/public-servers.html). | $2,000 |
| 3 | Sign Anything | Enable users to sign any transaction or message. | $3,000 |
| 4 | Sign and Submit Transaction | Enable users to sign and submit any transaction (including serialization). | $3,000 |
| 5 | Send Requests | Enable users to submit any rippled- or Clio-supported RPC request. | $??? |
| 5 | Dapp Integration | Offer APIs for dapps to interact with the Snap, with API documentation and snippets of example code. As a reference, see the [API to integrate BitcoinSnap into your dapp](https://github.com/KeystoneHQ/btcsnap/tree/master/packages/snap). | $8,000 |
| 6 | Ethereum Feature Parity | Add features similar or comparable to those available when using MetaMask with Ethereum. | Up to $50,000 |
| 7 | Sample App | Build a sample app to showcase how to interact with the Dapp Integration milestone. | $5,000 |
```
A couple of notes that I couldn't quite figure out how to word:
* I think 5 should be focused more on documentation and examples.
* I'm not sure what exactly you mean for 6. Snaps don't really support everything that exists in the Ethereum ecosystem yet - e.g. you can't add tokens to the dropdown, you can't use the normal Metamask UI at all yet. Perhaps that requirement should be added if/when Metamask allows Snaps to interact with the normal Metamask UI.

File: bounties/MetaMask Snap for the XRP Ledger.md [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+## Milestones
+
+| # | Description | Details | Proposed Potential Award |
+| - | ----------- | ------- | ------------------------ |
+| 1 | Create Account | Implement a Snap that securely generates and stores a new keypair and displays the associated XRPL address to the user. Make it easy for users to copy the address to share or use in other apps. | $1,000 |
+| 2 | Send XRP | Enable users to send XRP. The Snap should sign and submit the transaction to a server of the user's choice. For convenience, users can select from one of the [well-known public servers](https://xrpl.org/public-servers.html). | $2,000 |
+| 3 | Sign Anything | Enable users to sign any transaction or message. | $3,000 |
+| 4 | Sign and Submit Transaction | Enable users to sign and submit any transaction (including serialization). | $3,000 |
+| 5 | Send Requests | Enable users to submit any rippled- or Clio-supported RPC request. | $??? |
```
Comment (jonathanlei) [inline bounties/MetaMask Snap for the XRP Ledger.md:31]:
```suggestion
| 5 | Send Requests | Enable users to submit any rippled- or Clio-supported RPC request. | $3000 |
```

**Review by jonathanlei — APPROVED** [review]

File: bounties/MetaMask Snap for the XRP Ledger.md [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+| 3 | Sign Anything | Enable users to sign any transaction or message. | $3,000 |
+| 4 | Sign and Submit Transaction | Enable users to sign and submit any transaction (including serialization). | $3,000 |
+| 5 | Send Requests | Enable users to submit any rippled- or Clio-supported RPC request. | $3000 |
+| 6 | Documentation and Examples for Dapps | Enable easy [dapp](https://en.wikipedia.org/wiki/Decentralized_application) integrations: implement and document APIs for dapps to interact with the Snap, including example code. As a reference, see the [API to integrate BitcoinSnap into your dapp](https://github.com/KeystoneHQ/btcsnap/tree/master/packages/snap). | $8,000 |
+| 7 | Ethereum Feature Parity | Add features similar or comparable to those available when using MetaMask with Ethereum. While Snaps may not be able to use the normal MetaMask UI, you can implement support for tokens and advanced "smart transactors" by designing and building a custom UI. | Up to $50,000 |
+| 8 | Sample App | Build a sample app to showcase how to interact with the Dapp Integration milestone. | $5,000 |
```
Comment (JST5000) [inline bounties/MetaMask Snap for the XRP Ledger.md:34]: I'd recommend changing the order so related milestones are next to each other.
```suggestion
| 7 | Sample App | Build a sample app to showcase how to interact with the Dapp Integration milestone. (Milestone 6) | $5,000 |
| 8 | Ethereum Feature Parity | Add features similar or comparable to those available when using MetaMask with Ethereum. While Snaps may not be able to use the normal MetaMask UI, you can implement support for tokens and advanced "smart transactors" by designing and building a custom UI. | Up to $50,000 |
```

File: bounties/MetaMask Snap for the XRP Ledger.md [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+## Milestones
+
+| # | Description | Details | Proposed Potential Award |
+| - | ----------- | ------- | ------------------------ |
+| 1 | Create Account | Implement a Snap that securely generates and stores a new keypair and displays the associated XRPL address to the user. Make it easy for users to copy the address to share or use in other apps. | $1,000 |
+| 2 | Send XRP | Enable users to send XRP. The Snap should sign and submit the transaction to a server of the user's choice. For convenience, users can select from one of the [well-known public servers](https://xrpl.org/public-servers.html). | $2,000 |
+| 3 | Sign Anything | Enable users to sign any transaction or message. | $3,000 |
+| 4 | Sign and Submit Transaction | Enable users to sign and submit any transaction (including serialization). | $3,000 |
+| 5 | Send Requests | Enable users to submit any rippled- or Clio-supported RPC request. | $3000 |
```
Comment (develoQ) [inline bounties/MetaMask Snap for the XRP Ledger.md:31]: It's small, but I noticed it and will leave it as a comment.

```suggestion
| 5 | Send Requests | Enable users to submit any rippled- or Clio-supported RPC request. | $3,000 |
```

Comment (JST5000) [reply bounties/MetaMask Snap for the XRP Ledger.md]: Thanks! :)

File: bounties/MetaMask Snap for the XRP Ledger.md [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+| 3 | Sign Anything | Enable users to sign any transaction or message. | $3,000 |
+| 4 | Sign and Submit Transaction | Enable users to sign and submit any transaction (including serialization). | $3,000 |
+| 5 | Send Requests | Enable users to submit any rippled- or Clio-supported RPC request. | $3,000 |
+| 6 | Documentation and Examples for Dapps | Enable easy [dapp](https://en.wikipedia.org/wiki/Decentralized_application) integrations: implement and document APIs for dapps to interact with the Snap, including example code. As a reference, see the [API to integrate BitcoinSnap into your dapp](https://github.com/KeystoneHQ/btcsnap/tree/master/packages/snap). | $8,000 |
+| 7 | Sample App | Build a sample app to showcase how to interact with the Dapp Integration milestone. (Milestone 6) | $5,000 |
+| 8 | Ethereum Feature Parity | Add features similar or comparable to those available when using MetaMask with Ethereum. While Snaps may not be able to use the normal MetaMask UI, you can implement support for tokens and advanced "smart transactors" by designing and building a custom UI. | Up to $50,000 |
```
Comment (JST5000) [inline bounties/MetaMask Snap for the XRP Ledger.md:34]:
```suggestion
| 8 | Ethereum Feature Parity | Add features similar or comparable to those available when using MetaMask with Ethereum. While Snaps may not be able to use the normal MetaMask UI, you can implement support for tokens and advanced "smart transactors" by designing and building a custom UI. Feel free to propose ideas in the discussion corresponding to this milestone if you'd like feedback before building. | Up to $50,000 |
```

**Review by JST5000 — APPROVED** [review]
LGTM

**Review by ckniffen — APPROVED** [review]

**Review by mvadari — APPROVED** [review]

**Pull request merged**